(Brokered)Message refactoring

jtaubensee opened the pull request:
## Description
Addresses #98
BrokeredMessage refactoring:
* Changed BrokeredMessage -> Message and simplified the object.
  * Removed stream/serialization functionality
  * Body is an ArraySegment<byte>
  * Moved "Service Bus properties" to SystemProperties
  * Moved "user properties" to UserProperties
  * Removed `MessageIdGenerator`. To be added as a plugin (@seanfeldman)
* Updated dependencies

This checklist is used to make sure that common guidelines for a pull request are followed.

- [X] **I have read the [contribution guidelines](./CONTRIBUTING.md).**
- [X] Title of the pull request is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR.
- [X] The pull request does not introduce breaking changes (unless a major version change occurs in the assembly and module).
- [ ] If applicable, the public code is properly documented.
- [ ] Pull request includes test coverage for the included changes.
- [X] The code builds without any errors.

msftclas commented:
@jtaubensee,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

jtaubensee commented:
## Items still needing work:
* Message.Size: Need to provide support for body / size calculation
* Additional unit tests for Message and AMQPConverter

jtaubensee commented:
![passing tests](https://cloud.githubusercontent.com/assets/8162917/24181267/2b70b8b0-0e77-11e7-8d71-8b844ab1d9c4.png)

jtaubensee commented on src/Microsoft.Azure.ServiceBus/Message.cs:
```diff
@@ -6,1136 +6,200 @@ namespace Microsoft.Azure.ServiceBus
-            get
-            {
-                this.ThrowIfDisposed();
-                this.ThrowIfNotReceived();
```
Should this be removed? i.e. Should the default behavior be to throw when trying to access a property that would only be set when a message is received?

jtaubensee replied:
@SeanFeldman / @nemakam / @vinaysurya - Any thoughts here?

SeanFeldman replied:
Makes sense to keep it. If message was never received (ie locally created message), `DeliveryCount` is not valid. Alternative (if you don't want to throw) would be to return an invalid value (-1?). I'm not sure what would be better. 

Side note: I've added `.IsReceived` on the message for (PR #78) to avoid exceptions when dealing with messages w/o the knowing if it was received or created locally.

nemakam replied:
I'm also in favor of throwing here. Different data types will have different 'invalid value', and might make it tricky. Throwing exception in conjunction with 'IsReceived' property sounds like a good approach

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Message.cs:
```diff
@@ -6,1136 +6,200 @@ namespace Microsoft.Azure.ServiceBus
         /// </exception>
         /// <param name="messageId"> Identifier for the message. </param>
-        static void ValidateMessageId(string messageId)
+        private void ValidateMessageId(string messageId)
```
Any specific reason why this shouldn't be static?

jtaubensee replied:
Does it need to be static? Its a private method...

SeanFeldman replied:
For performance. A single static method as it doesn't need to access any non-static members.

jtaubensee replied:
https://msdn.microsoft.com/library/79b3xss3(v=vs.110).aspx tldr;
>A call to a static method generates a call instruction in Microsoft intermediate language (MSIL), whereas a call to an instance method generates a callvirt instruction, which also checks for a null object references. However, **most of the time** the performance difference between the two is **not significant**.

SeanFeldman replied:
While you might not get a lot of perf, it is on the hot path (executed per each message going out). If there's no state, really no reason to keep it non-static.

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Message.cs:
```diff
@@ -1150,7 +214,7 @@ static void ValidateMessageId(string messageId)
         /// </exception>
         /// <param name="sessionId"> Identifier for the session. </param>
-        static void ValidateSessionId(string sessionId)
+        private void ValidateSessionId(string sessionIdPropertyName, string sessionId)
```
Why not static?

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Message.cs:
```diff
@@ -1159,7 +223,7 @@ static void ValidateSessionId(string sessionId)
         }
 
-        static void ValidatePartitionKey(string partitionKeyPropertyName, string partitionKey)
+        private void ValidatePartitionKey(string partitionKeyPropertyName, string partitionKey)
```
Why not static?

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Message.cs:
```diff
@@ -1168,234 +232,26 @@ static void ValidatePartitionKey(string partitionKeyPropertyName, string partiti
-        /// <summary> Throw if not locked. </summary>
-        /// <exception cref="FxTrace.Exception"> Thrown when as error. </exception>
-        void ThrowIfNotLocked()
+        public sealed class SystemPropertiesCollection
```
It says collection, but it's more like `SystemProperties`.

ApolloCreek replied:
… if so, then embed the class in its only using class. Otherwise move it to an own file!

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Message.cs:
```diff
@@ -1168,234 +232,26 @@ static void ValidatePartitionKey(string partitionKeyPropertyName, string partiti
-        /// <summary> Throw if not locked. </summary>
-        /// <exception cref="FxTrace.Exception"> Thrown when as error. </exception>
-        void ThrowIfNotLocked()
+        public sealed class SystemPropertiesCollection
```
Deserves a file of its own? 😃

nemakam replied:
@SeanFeldman , do you see any particular concern with keeping the class definition in the same file? This class is never used outside of Message context and should not be.

SeanFeldman replied:
It's much easier to navigate to the class when the filename represent the class and not a container class. Not as critical as the name comment.

nemakam replied:
@SeanFeldman, true for the navigation. But with advent of IDEs, I think it shouldn't be that big of a concern. In contrast, if you do not split the file, then you have a logical closure on where the class is being used. I personally would prefer having it in same class in this scenario when it strictly doesn't make sense to be used outside the scope of a Message

SeanFeldman replied:
For classes that are not public, perhaps. For public stuff I'd push back. The closure can be provided by the the IDEs and tooling such as VS2017 and R#. GH PRs are often reviewed in the browser. Navigation by file type is allowed on all branches. Search is only on the active branch. So navigating to `SystemPropertiesCollection` would not be possible. But again, this is not critical, just a suggestion. Name is more important.

jtaubensee replied:
So if we keep it in the same file and rename the class to just `SystemProperties` there is a name clash between that and the property on `Message`. I am of the opinion to keep the class in the file, and therefore the name would stay the same. I personally like the clarity of 'this class is only ever used in here'

SeanFeldman commented on test/Microsoft.Azure.ServiceBus.UnitTests/AmqpConverterTests.cs:
```diff
@@ -0,0 +1,68 @@
+
+    public class AmqpConverterTests
+    {
+        public class WhenConvertingBrokeredMessagesToAmqpMessages
```
`WhenConvertingBrokeredMessagesToAmqpMessages` vs `When_converting_brokered_messages_to_amqp_messages` I'll take 2nd option any time as it's easier to read. If this was a German word, that would be a different story 😉

SeanFeldman commented on samples/ReceiveSample/Program.cs:
```diff
@@ -40,7 +41,7 @@ private static void ReceiveMessages()
                     {
                         // Process the message
-                        Console.WriteLine($"Received message: SequenceNumber:{message.SequenceNumber} Body:{message.GetBody<string>()}");
+                        Console.WriteLine($"Received message: SequenceNumber:{message.SystemProperties.SequenceNumber} Body:{Encoding.UTF8.GetString(message.Body.Array)}");
```
`message.Body.Array` feels like friction.

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Message.cs:
```diff
@@ -6,1136 +6,200 @@ namespace Microsoft.Azure.ServiceBus
-        /// <exception cref="ObjectDisposedException">Thrown if message is in disposed state.</exception>
-        /// <exception cref="InvalidOperationException">Thrown if <seealso cref="PartitionKey"/> or <seealso cref="SessionId"/> are set to different values.</exception>
-        internal string Publisher
+        public ArraySegment<byte> Body { get; set; }
```
Any specific reason why `Body` is of type `ArraySegment<byte>` and not `byte[]`? Manipulations would be done on the entire body, not subsets of bytes, so what is the benefit?

jtaubensee replied:
The idea is that if you need to split up a larger message into multiple smaller messages, this would be simpler. Followed guidance [here](https://msdn.microsoft.com/library/1hsbd92d(v=vs.110).aspx#Remarks). I think that ArraySegment is probably OK, but since a lot of customers deal with strings, we should provide and overload for that, and an easy way to access the body as a string.

SeanFeldman replied:
So that makes sense in the context of the application that is handling the body as a source of multiple messages. For the general purpose, you could return `byte[]`. 

For the case where you need to operate on sections of the body it would be `new ArraySegment<byte>(msg.Body)`. Why to impose `ArraySegment<byte>` on all use cases?

SeanFeldman commented:
Are we going back to the `BrokeredMessage`?

jtaubensee commented:
@SeanFeldman - I don't think we are going back to BrokeredMessage, I just used the alias to make it clear what the AMQP converter was doing. I can just as easily use an alias like "SBMessage" in that class too.

jtaubensee commented on src/Microsoft.Azure.ServiceBus/Message.cs:
```diff
@@ -6,1136 +6,200 @@ namespace Microsoft.Azure.ServiceBus
 
-    /// <summary>Represents the unit of communication between ServiceBus client and Service.</summary>
-    public sealed class Message : IDisposable
+    public class Message
```
per @vinaysurya this should still implement IDisposible. How about being sealed?

nemakam commented on src/Microsoft.Azure.ServiceBus/Message.cs:
```diff
@@ -6,1136 +6,200 @@ namespace Microsoft.Azure.ServiceBus
-                    throw new InvalidOperationException("BrokeredMessageStreamNotCloneable");
-                }
+                var clonedBody = new byte[this.Body.Count];
+                Array.Copy(this.Body.Array, clonedBody, this.Body.Count);
```
This doesn't copy as intended.
ArraySegment has an offset and a count. The offset need not be 0.

nemakam replied:
Can you please add a test which initializes a message with a true arraySegment (part of byte[] which neither starts or ends at the start or end of the byte[]). You need to test for send, receive and a clone for that.

SeanFeldman replied:
@nemakam you mean `Array.Copy(src.Array, src.Offset, dst, 0, src.Count);`?
I'm still not convinced this is needed... if you send a giant payload and chunk it, it's the application responsibility, not the brokered message. Brokered message should just accept a `byte[]`.

nemakam replied:
I'd also personally prefer using byte[]. But this comment was keeping in mind if ArraySegment is chosen. And yes, your overload of Array.Copy looks right.

SeanFeldman replied:
It feels like `ArraySegment<byte>` is more of an application concern, not the brokered message. Brokered message should deliver the bytes and the app can choose how to process/operate on those.

jtaubensee commented:
Latest set of changes:
- Changed `Message.Body` from `ArraySegment<byte>` to `byte[]`
- Added `Message.BodyAsString` method which provides the byte[] body as a UTF8 encoded string
- Moved `Message.LockToken` to `Message.SystemProperties.LockToken`
- Changed `BrokeredMessage` alias inside of `AmqpConverter` to `SBMessage` (letting BrokeredMessage rest peacefully :smile:).

SeanFeldman commented on src/Microsoft.Azure.ServiceBus/Message.cs:
```diff
@@ -6,1136 +6,202 @@ namespace Microsoft.Azure.ServiceBus
+        public Message(byte[] array)
+        {
+            this.Body = array;
+            this.SystemProperties = new SystemPropertiesCollection();
```
Collection implies a set of item of the same time, where here it's just a group of properties. Should it be just a `SystemProperties` type?

jtaubensee replied:
So if we keep it in the same file and rename the class to just SystemProperties there is a name clash between that and the property on Message. I am of the opinion to keep the class in the file, and therefore the name would stay the same. I personally like the clarity of 'this class is only ever used in here'

SeanFeldman replied:
@jtaubensee not following... there's no clash.
If you have 2 files (`Message.cs` and `SystemProperties.cs`) or not, it is valid for `Message` class to contain a property of `SystemProperties` type and be named `SystemProperties`.

```c#
    public class SystemProperties
    {
        public int SequenceNumber { get; set; }
    }
```

```c#
    public class Message
    {
        public SystemProperties SystemProperties { get; set; }
    }
```

jtaubensee replied:
This happens when `SystemProperties` is in the same class:

![2017-03-23_10h58_51](https://cloud.githubusercontent.com/assets/8162917/24262597/c79f8270-0fb7-11e7-9a21-7c18c0fec176.png)

jtaubensee replied:
Based on my earlier comment, I'd prefer to leave it in the same file. I think it is more clear that the `SystemProperties` class is used specifically in `Message`.

SeanFeldman replied:
I'm not sure why you get that error... this is what I see in a quick spike:
![image](https://cloud.githubusercontent.com/assets/1309622/24265010/bb035928-0fc7-11e7-9fd4-4800699db4f7.png)

nemakam replied:
The difference would be Innerclass vs "two classes in the same file"...

SeanFeldman replied:
Why would `SystemProperties` be an inner class of `Message` if it's public I do not follow...anyhow let it be suffixed with `Collection` for fun 😃

nemakam replied:
@SeanFeldman, Just found this online: https://coderanch.com/t/579241/java/public-nested-classes-good-practice. 
I'm sure there might be better articles than this online...

SeanFeldman replied:
@nemakam not sure. All depends on the context.

jtaubensee commented:
Created #109 to track Message.Size functionality

SeanFeldman commented:
@jtaubensee should this be tracked separately as well?

> https://github.com/MessageHandler/MessageHandler.Fabric/pull/2

niemyjski commented:
Do we even need (Message.BodyAsString). I'd rather keep this stuff really clean, if you want it as a string you could easily do that yourself.

nemakam commented:
@niemyjski , I would still prefer having the string overload for set-get of Body as sending string messages is one of the most common use-cases. I'm sure its a do-able thing for the customer himself, but having the overload makes it easier and makes the code look better, as compared to doing `Encoding.UTF8.GetString()` everytime.

niemyjski commented:
Is it lazy loaded? or do we have to take that allocation hit every time even if we don't need or want it?

nemakam commented:
@niemyjski. it is being implemented as `public string BodyAsString => Encoding.UTF8.GetString(this.Body);`. This should be lazy loaded.

niemyjski commented:
@nemakam shouldn't this be stored in a private member the first time you access it? If you call that multiple times that's a lot of work each time you call BodyAsString

SeanFeldman commented:
@niemyjski why would you application call this property multiple times?

Assuming your code is using `BodyAsString` multiple times, I'd counter argue with the following: if you create a message locally, and modify it, having a cached body would fire back. To address that, you'd need to introduce a staleness check, which makes the whole thing way more complicated than it should be.

niemyjski commented:
Your message is immutable and shouldn't ever change. I've seen sample code as well as code in prod where bodyAsString/similar property is called multiple times without realizing what is happening under the covers.

SeanFeldman commented:
Personally, I think this `BodyAsAnything` isn't necessary since `byte[]` is all I care for. There's no `BodyAsStream` or other variant, so what's the point? KISS principle 😃

niemyjski commented:
@SeanFeldman I agree!

jtaubensee commented:
@niemyjski / @SeanFeldman - Sorry for the delay, I was out of the office last week. To clarify, the body is definitely mutable with this client, however, I dont have any problem removing the string over load functionality for Message.Body, as we can gauge user feedback and easily add it at anytime.

SeanFeldman commented:
@jtaubensee agreed. If it's a needed feature, people will ask/vote for it.
